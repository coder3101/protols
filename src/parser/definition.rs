use async_lsp::lsp_types::{Location, Range};
use tree_sitter::Node;

use crate::{nodekind::NodeKind, utils::ts_to_lsp_position};

use super::ParsedTree;

impl ParsedTree {
    pub fn definition(&self, identifier: &str, content: impl AsRef<[u8]>) -> Vec<Location> {
        let mut results = vec![];
        self.definition_impl(identifier, self.tree.root_node(), &mut results, content);
        results
    }
    fn definition_impl(
        &self,
        identifier: &str,
        n: Node,
        v: &mut Vec<Location>,
        content: impl AsRef<[u8]>,
    ) {
        if identifier.is_empty() {
            return;
        }

        if !identifier.contains('.') {
            let locations: Vec<Location> = self
                .filter_nodes_from(n, NodeKind::is_userdefined)
                .into_iter()
                .filter(|n| n.utf8_text(content.as_ref()).expect("utf-8 parse error") == identifier)
                .map(|n| Location {
                    uri: self.uri.clone(),
                    range: Range {
                        start: ts_to_lsp_position(&n.start_position()),
                        end: ts_to_lsp_position(&n.end_position()),
                    },
                })
                .collect();

            v.extend(locations);
            return;
        }

        // Safety: identifier contains a .
        let (parent_identifier, remaining) = identifier.split_once('.').unwrap();
        let child_node = self
            .filter_nodes_from(n, NodeKind::is_userdefined)
            .into_iter()
            .find(|n| n.utf8_text(content.as_ref()).expect("utf8-parse error") == parent_identifier)
            .and_then(|n| n.parent());

        if let Some(inner) = child_node {
            self.definition_impl(remaining, inner, v, content);
        }
    }
}

#[cfg(test)]
mod test {
    use async_lsp::lsp_types::Url;
    use insta::assert_yaml_snapshot;

    use crate::parser::ProtoParser;

    #[test]
    fn test_goto_definition() {
        let url: Url = "file://foo/bar.proto".parse().unwrap();
        let contents = include_str!("input/test_goto_definition.proto");
        let parsed = ProtoParser::new().parse(url, contents);

        assert!(parsed.is_some());
        let tree = parsed.unwrap();
        assert_yaml_snapshot!(tree.definition("Author", contents));
        assert_yaml_snapshot!(tree.definition("", contents));
    }
}
