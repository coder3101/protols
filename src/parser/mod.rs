use async_lsp::lsp_types::Url;
use tree_sitter::Tree;

mod definition;
mod diagnostics;
mod docsymbol;
mod hover;
mod rename;
mod tree;

pub struct ProtoParser {
    parser: tree_sitter::Parser,
}

pub struct ParsedTree {
    pub uri: Url,
    tree: Tree,
}

impl ProtoParser {
    pub fn new() -> Self {
        let mut parser = tree_sitter::Parser::new();
        if let Err(e) = parser.set_language(&protols_tree_sitter_proto::language()) {
            panic!("failed to set ts language parser {:?}", e);
        }
        Self { parser }
    }

    pub fn parse(&mut self, uri: Url, contents: impl AsRef<[u8]>) -> Option<ParsedTree> {
        self.parser
            .parse(contents, None)
            .map(|t| ParsedTree { tree: t, uri })
    }
}
