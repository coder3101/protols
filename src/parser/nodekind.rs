use async_lsp::lsp_types::SymbolKind;
use tree_sitter::Node;

#[allow(unused)]
pub enum NodeKind {
    Identifier,
    Error,
    MessageName,
    EnumName,
    FieldName,
    ServiceName,
    RpcName,
    PackageName,
}

#[allow(unused)]
impl NodeKind {
    pub fn as_str(&self) -> &'static str {
        match self {
            NodeKind::Identifier => "identifier",
            NodeKind::Error => "ERROR",
            NodeKind::MessageName => "message_name",
            NodeKind::EnumName => "enum_name",
            NodeKind::FieldName => "message_or_enum_type",
            NodeKind::ServiceName => "service_name",
            NodeKind::RpcName => "rpc_name",
            NodeKind::PackageName => "package_name",
        }
    }

    pub fn is_identifier(n: &Node) -> bool {
        n.kind() == "identifier"
    }

    pub fn is_error(n: &Node) -> bool {
        n.kind() == "ERROR"
    }

    pub fn is_userdefined(n: &Node) -> bool {
        matches!(n.kind(), "message_name" | "enum_name")
    }

    pub fn is_actionable(n: &Node) -> bool {
        matches!(
            n.kind(),
            "message_name" | "enum_name" | "message_or_enum_type" | "rpc_name" | "service_name"
        )
    }

    pub fn to_symbolkind(n: &Node) -> SymbolKind {
        match n.kind() {
            "message_name" => SymbolKind::STRUCT,
            "enum_name" => SymbolKind::ENUM,
            _ => SymbolKind::NULL,
        }
    }
}


