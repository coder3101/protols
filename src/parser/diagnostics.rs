use async_lsp::lsp_types::{Diagnostic, DiagnosticSeverity, PublishDiagnosticsParams, Range};

use crate::{nodekind::NodeKind, utils::ts_to_lsp_position};

use super::ParsedTree;

impl ParsedTree {
    pub fn collect_parse_errors(&self) -> PublishDiagnosticsParams {
        let diagnostics = self
            .filter_nodes(NodeKind::is_error)
            .into_iter()
            .map(|n| Diagnostic {
                range: Range {
                    start: ts_to_lsp_position(&n.start_position()),
                    end: ts_to_lsp_position(&n.end_position()),
                },
                severity: Some(DiagnosticSeverity::ERROR),
                source: Some("protols".to_string()),
                message: "Syntax error".to_string(),
                ..Default::default()
            })
            .collect();
        PublishDiagnosticsParams {
            uri: self.uri.clone(),
            diagnostics,
            version: None,
        }
    }
}

#[cfg(test)]
mod test {
    use async_lsp::lsp_types::Url;
    use insta::assert_yaml_snapshot;

    use crate::parser::ProtoParser;

    #[test]
    fn test_collect_parse_error() {
        let url: Url = "file://foo/bar.proto".parse().unwrap();
        let contents = include_str!("input/test_collect_parse_error1.proto");

        let parsed = ProtoParser::new().parse(url.clone(), contents);
        assert!(parsed.is_some());
        assert_yaml_snapshot!(parsed.unwrap().collect_parse_errors());

        let contents = include_str!("input/test_collect_parse_error2.proto");

        let parsed = ProtoParser::new().parse(url.clone(), contents);
        assert!(parsed.is_some());
        assert_yaml_snapshot!(parsed.unwrap().collect_parse_errors());
    }
}
