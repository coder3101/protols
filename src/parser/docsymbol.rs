use async_lsp::lsp_types::{DocumentSymbol, Range};
use tree_sitter::TreeCursor;

use crate::{nodekind::NodeKind, utils::ts_to_lsp_position};

use super::ParsedTree;

#[derive(Default)]
pub(super) struct DocumentSymbolTreeBuilder {
    // The stack are things we're still in the process of building/parsing.
    stack: Vec<(usize, DocumentSymbol)>,
    // The found are things we've finished processing/parsing, at the top level of the stack.
    found: Vec<DocumentSymbol>,
}

impl DocumentSymbolTreeBuilder {
    pub(super) fn push(&mut self, node: usize, symbol: DocumentSymbol) {
        self.stack.push((node, symbol));
    }

    pub(super) fn maybe_pop(&mut self, node: usize) {
        let should_pop = self.stack.last().map_or(false, |(n, _)| *n == node);
        if should_pop {
            let (_, explored) = self.stack.pop().unwrap();
            if let Some((_, parent)) = self.stack.last_mut() {
                parent.children.as_mut().unwrap().push(explored);
            } else {
                self.found.push(explored);
            }
        }
    }

    pub(super) fn build(self) -> Vec<DocumentSymbol> {
        self.found
    }
}

impl ParsedTree {
    pub fn find_document_locations(&self, content: impl AsRef<[u8]>) -> Vec<DocumentSymbol> {
        let mut builder = DocumentSymbolTreeBuilder::default();
        let content = content.as_ref();

        let mut cursor = self.tree.root_node().walk();
        self.find_document_locations_inner(&mut builder, &mut cursor, content);

        builder.build()
    }

    fn find_document_locations_inner(
        &self,
        builder: &mut DocumentSymbolTreeBuilder,
        cursor: &'_ mut TreeCursor,
        content: &[u8],
    ) {
        loop {
            let node = cursor.node();

            if NodeKind::is_userdefined(&node) {
                let name = node.utf8_text(content).unwrap();
                let kind = NodeKind::to_symbolkind(&node);
                let detail = self.find_preceding_comments(node.id(), content);
                let message = node.parent().unwrap();

                // https://github.com/rust-lang/rust/issues/102777
                #[allow(deprecated)]
                let new_symbol = DocumentSymbol {
                    name: name.to_string(),
                    detail,
                    kind,
                    tags: None,
                    deprecated: None,
                    range: Range {
                        start: ts_to_lsp_position(&message.start_position()),
                        end: ts_to_lsp_position(&message.end_position()),
                    },
                    selection_range: Range {
                        start: ts_to_lsp_position(&node.start_position()),
                        end: ts_to_lsp_position(&node.end_position()),
                    },
                    children: Some(vec![]),
                };

                builder.push(message.id(), new_symbol);
            }

            if cursor.goto_first_child() {
                self.find_document_locations_inner(builder, cursor, content);
                builder.maybe_pop(node.id());
                cursor.goto_parent();
            }

            if !cursor.goto_next_sibling() {
                break;
            }
        }
    }
}

#[cfg(test)]
mod test {
    use async_lsp::lsp_types::Url;
    use insta::assert_yaml_snapshot;

    use crate::parser::ProtoParser;

    #[test]
    #[allow(deprecated)]
    fn test_document_symbols() {
        let uri: Url = "file://foo/bar/pro.proto".parse().unwrap();
        let contents = include_str!("input/test_document_symbols.proto");

        let parsed = ProtoParser::new().parse(uri.clone(), contents);
        assert!(parsed.is_some());

        let tree = parsed.unwrap();
        assert_yaml_snapshot!(tree.find_document_locations(contents));
    }
}
