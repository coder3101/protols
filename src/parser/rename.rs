use std::collections::HashMap;

use async_lsp::lsp_types::{Position, Range, TextEdit, WorkspaceEdit};

use crate::{nodekind::NodeKind, utils::ts_to_lsp_position};

use super::ParsedTree;

impl ParsedTree {
    pub fn can_rename(&self, pos: &Position) -> Option<Range> {
        self.get_node_at_position(pos)
            .filter(NodeKind::is_identifier)
            .and_then(|n| {
                if n.parent().is_some() && NodeKind::is_userdefined(&n.parent().unwrap()) {
                    Some(Range {
                        start: ts_to_lsp_position(&n.start_position()),
                        end: ts_to_lsp_position(&n.end_position()),
                    })
                } else {
                    None
                }
            })
    }

    pub fn rename(
        &self,
        pos: &Position,
        new_text: &str,
        content: impl AsRef<[u8]>,
    ) -> Option<WorkspaceEdit> {
        let old_text = self
            .get_node_text_at_position(pos, content.as_ref())
            .unwrap_or_default();

        let mut changes = HashMap::new();

        let diff: Vec<_> = self
            .filter_nodes(NodeKind::is_identifier)
            .into_iter()
            .filter(|n| n.utf8_text(content.as_ref()).unwrap() == old_text)
            .map(|n| TextEdit {
                new_text: new_text.to_string(),
                range: Range {
                    start: ts_to_lsp_position(&n.start_position()),
                    end: ts_to_lsp_position(&n.end_position()),
                },
            })
            .collect();

        if diff.is_empty() {
            return None;
        }

        changes.insert(self.uri.clone(), diff);

        Some(WorkspaceEdit {
            changes: Some(changes),
            ..Default::default()
        })
    }
}

#[cfg(test)]
mod test {
    use async_lsp::lsp_types::{Position, Url};
    use insta::assert_yaml_snapshot;

    use crate::parser::ProtoParser;

    #[test]
    fn test_rename() {
        let uri: Url = "file://foo/bar.proto".parse().unwrap();
        let pos_book_rename = Position {
            line: 5,
            character: 9,
        };
        let pos_author_rename = Position {
            line: 21,
            character: 10,
        };
        let pos_non_renamble = Position {
            line: 24,
            character: 4,
        };
        let contents = include_str!("input/test_rename.proto");

        let parsed = ProtoParser::new().parse(uri.clone(), contents);
        assert!(parsed.is_some());
        let tree = parsed.unwrap();

        assert_yaml_snapshot!(tree.rename(&pos_book_rename, "Kitab", contents));
        assert_yaml_snapshot!(tree.rename(&pos_author_rename, "Writer", contents));
        assert_yaml_snapshot!(tree.rename(&pos_non_renamble, "Doesn't matter", contents));
    }

    #[test]
    fn test_can_rename() {
        let uri: Url = "file://foo/bar/test.proto".parse().unwrap();
        let pos_rename = Position {
            line: 5,
            character: 9,
        };
        let pos_non_rename = Position {
            line: 2,
            character: 2,
        };
        let pos_inner_type = Position {
            line: 19,
            character: 11,
        };
        let pos_outer_type = Position {
            line: 19,
            character: 5,
        };

        let contents = include_str!("input/test_can_rename.proto");
        let parsed = ProtoParser::new().parse(uri.clone(), contents);
        assert!(parsed.is_some());

        let tree = parsed.unwrap();
        assert_yaml_snapshot!(tree.can_rename(&pos_rename));
        assert_yaml_snapshot!(tree.can_rename(&pos_non_rename));
        assert_yaml_snapshot!(tree.can_rename(&pos_inner_type));
        assert_yaml_snapshot!(tree.can_rename(&pos_outer_type));
    }
}
