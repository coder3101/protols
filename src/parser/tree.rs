use async_lsp::lsp_types::Position;
use tree_sitter::{Node, TreeCursor};

use crate::utils::lsp_to_ts_point;

use super::{nodekind::NodeKind, ParsedTree};

impl ParsedTree {
    pub(super) fn walk_and_collect_filter<'a>(
        cursor: &mut TreeCursor<'a>,
        f: fn(&Node) -> bool,
    ) -> Vec<Node<'a>> {
        let mut v = vec![];

        loop {
            let node = cursor.node();

            if f(&node) {
                v.push(node)
            }

            if cursor.goto_first_child() {
                v.extend(Self::walk_and_collect_filter(cursor, f));
                cursor.goto_parent();
            }

            if !cursor.goto_next_sibling() {
                break;
            }
        }

        v
    }

    pub(super) fn advance_cursor_to(cursor: &mut TreeCursor<'_>, nid: usize) -> bool {
        loop {
            let node = cursor.node();
            if node.id() == nid {
                return true;
            }
            if cursor.goto_first_child() {
                if Self::advance_cursor_to(cursor, nid) {
                    return true;
                }
                cursor.goto_parent();
            }
            if !cursor.goto_next_sibling() {
                return false;
            }
        }
    }

    pub(super) fn get_node_text_at_position<'a>(
        &'a self,
        pos: &Position,
        content: &'a [u8],
    ) -> Option<&'a str> {
        self.get_node_at_position(pos)
            .map(|n| n.utf8_text(content.as_ref()).expect("utf-8 parse error"))
    }

    pub(super) fn get_actionable_node_text_at_position<'a>(
        &'a self,
        pos: &Position,
        content: &'a [u8],
    ) -> Option<&'a str> {
        self.get_actionable_node_at_position(pos)
            .map(|n| n.utf8_text(content.as_ref()).expect("utf-8 parse error"))
    }

    pub(super) fn get_actionable_node_at_position<'a>(
        &'a self,
        pos: &Position,
    ) -> Option<Node<'a>> {
        self.get_node_at_position(pos)
            .map(|n| {
                if NodeKind::is_actionable(&n) {
                    n
                } else {
                    n.parent().unwrap()
                }
            })
            .filter(NodeKind::is_actionable)
    }

    pub(super) fn get_node_at_position<'a>(&'a self, pos: &Position) -> Option<Node<'a>> {
        let pos = lsp_to_ts_point(pos);
        self.tree.root_node().descendant_for_point_range(pos, pos)
    }

    pub(super) fn filter_node(&self, f: fn(&Node) -> bool) -> Vec<Node> {
        let mut cursor = self.tree.root_node().walk();
        Self::walk_and_collect_filter(&mut cursor, f)
    }
}

#[cfg(test)]
mod test {
    use tree_sitter::Node;

    use crate::parser::ProtoParser;

    fn is_message(n: &Node) -> bool {
        n.kind() == "message_name"
    }

    #[test]
    fn test_find_children_by_kind() {
        let contents = r#"syntax = "proto3";

package com.book;

message Book {

    message Author {
        string name = 1;
        string country = 2;
    };
    // This is a multi line comment on the field name
    // Of a message called Book
    int64 isbn = 1;
    string title = 2;
    Author author = 3;
}
"#;
        let parsed = ProtoParser::new().parse(contents);
        assert!(parsed.is_some());
        let tree = parsed.unwrap();
        let nodes = tree.filter_node(is_message);

        assert_eq!(nodes.len(), 2);

        let names: Vec<_> = nodes
            .into_iter()
            .map(|n| n.utf8_text(contents.as_ref()).unwrap())
            .collect();
        assert_eq!(names[0], "Book");
        assert_eq!(names[1], "Author");
    }
}
