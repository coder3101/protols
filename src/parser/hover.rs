use async_lsp::lsp_types::MarkedString;
use tree_sitter::Node;

use crate::nodekind::NodeKind;

use super::ParsedTree;

impl ParsedTree {
    pub(super) fn find_preceding_comments(
        &self,
        nid: usize,
        content: impl AsRef<[u8]>,
    ) -> Option<String> {
        let root = self.tree.root_node();
        let mut cursor = root.walk();

        Self::advance_cursor_to(&mut cursor, nid);
        if !cursor.goto_parent() {
            return None;
        }

        if !cursor.goto_previous_sibling() {
            return None;
        }

        let mut comments = vec![];
        while cursor.node().kind() == "comment" {
            let node = cursor.node();
            let text = node
                .utf8_text(content.as_ref())
                .expect("utf-8 parser error")
                .trim()
                .trim_start_matches("//")
                .trim();

            comments.push(text);

            if !cursor.goto_previous_sibling() {
                break;
            }
        }
        if !comments.is_empty() {
            comments.reverse();
            Some(comments.join("\n"))
        } else {
            None
        }
    }

    pub fn hover(&self, identifier: &str, content: impl AsRef<[u8]>) -> Vec<MarkedString> {
        let mut results = vec![];
        self.hover_impl(identifier, self.tree.root_node(), &mut results, content);
        results
    }

    fn hover_impl(
        &self,
        identifier: &str,
        n: Node,
        v: &mut Vec<MarkedString>,
        content: impl AsRef<[u8]>,
    ) {
        if identifier.is_empty() {
            return;
        }

        if !identifier.contains(".") {
            let comments: Vec<MarkedString> = self
                .filter_nodes_from(n, NodeKind::is_userdefined)
                .into_iter()
                .filter(|n| n.utf8_text(content.as_ref()).expect("utf-8 parse error") == identifier)
                .filter_map(|n| self.find_preceding_comments(n.id(), content.as_ref()))
                .map(MarkedString::String)
                .collect();

            v.extend(comments);
            return;
        }

        // Safety: identifier contains a .
        let (parent_identifier, remaining) = identifier.split_once(".").unwrap();
        let child_node = self
            .filter_nodes_from(n, NodeKind::is_userdefined)
            .into_iter()
            .find(|n| n.utf8_text(content.as_ref()).expect("utf8-parse error") == parent_identifier)
            .and_then(|n| n.parent());

        if let Some(inner) = child_node {
            self.hover_impl(remaining, inner, v, content);
        }
    }
}

#[cfg(test)]
mod test {
    use async_lsp::lsp_types::Url;
    use insta::assert_yaml_snapshot;

    use crate::parser::ProtoParser;

    #[test]
    fn test_hover() {
        let uri: Url = "file://foo.bar/p.proto".parse().unwrap();
        let contents = include_str!("input/test_hover.proto");
        let parsed = ProtoParser::new().parse(uri.clone(), contents);

        assert!(parsed.is_some());
        let tree = parsed.unwrap();

        let res = tree.hover("Book", contents);
        assert_yaml_snapshot!(res);

        let res = tree.hover("", contents);
        assert_yaml_snapshot!(res);

        let res = tree.hover("Book.Author", contents);
        assert_yaml_snapshot!(res);

        let res = tree.hover("Comic.Author", contents);
        assert_yaml_snapshot!(res);

        let res = tree.hover("Author", contents);
        assert_yaml_snapshot!(res);
    }
}
