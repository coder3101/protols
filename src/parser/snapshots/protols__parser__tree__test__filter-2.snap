---
source: src/parser/tree.rs
expression: package_name
---
com.parser
